openpathsampling/pathmover.py

```diff
@@ -1244,6 +1286,146 @@ def move(self, globalstate):
+        the PathMovers to choose from
+    weights : list of floats
+        the relative weight of each PathMover (does not need to be normalized)
+    """
+
+    def _selector(self, globalstate):
+        if self.weights is None:
+            weights = [1.0] * len(self.movers)
+        else:
+            weights = self.weights
+
+        # this is implemented by setting all weights locally to zero that
+        # correspond to movers that will potentially fail since the required
+        # input ensembles are not present in the globalstate
+
+        present_ensembles = globalstate.ensembles
+
+        for idx, mover in enumerate(self.movers):
+            for ens in mover.input_ensembles:
+                if ens not in present_ensembles:
+                    # ens might be required but is not present
+                    weights[idx] = 0.0
+
+        return weights
+
+class FirstAllowedMover(RandomSelectionMover):
+    """
+    Chooses a first mover that has samples in all required ensembles.
+
+    A mover can only safely be run, if all inputs can be satisfied. This will pick
+    the first mover from the list where all ensembles from input_ensembles are
+    found.
+
+    Attributes
+    ----------
+    movers : list of PathMover
+        the PathMovers to choose from
+    """
+
+    def _selector(self, globalstate):
+        weights = [1.0] * len(self.movers)
+
+        present_ensembles = globalstate.ensembles
+
+        found = False
+
+        for idx, mover in enumerate(self.movers):
+            if not found:
+                for ens in mover.input_ensembles:
+                    if ens not in present_ensembles:
+                        # ens might be required but is not present
+                        weights[idx] = 0.0
+
+                if weights[idx] > 0.0:
+                    found = True
+            else:
+                weights[idx] = 0.0
+
+        return weights
+
+class LastAllowedMover(RandomSelectionMover):
+    """
+    Chooses the last mover that has samples in all required ensembles.
+
+    A mover can only safely be run, if all inputs can be satisfied. This will pick
+    the first mover from the list where all ensembles from input_ensembles are
+    found.
+
+    Attributes
+    ----------
+    movers : list of PathMover
+        the PathMovers to choose from
+    """
+
+    def _selector(self, globalstate):
+        weights = [1.0] * len(self.movers)
+
+        present_ensembles = globalstate.ensembles
+
+        found = False
+
+        for idx, mover in reversed(enumerate(self.movers)):
+            if not found:
+                for ens in mover.input_ensembles:
+                    if ens not in present_ensembles:
+                        # ens might be required but is not present
+                        weights[idx] = 0.0
+
+                if weights[idx] > 0.0:
+                    found = True
+            else:
+                weights[idx] = 0.0
+
+        return weights
```

Comment:
(starting with a quote from the end, because the answer to that answers a lot)

> I hope you did not think that I wanted it to be this way from the start. It was my attempt to combine your bootstrapping with the movers in a consistent way.

No, I didn't think that at all. What I meant to say is that, if I recall correctly, It was ***me*** who started this direction. ***I*** started us down this road when I introduced the `PartialAcceptanceMover`. (As much emphasis as possible, to say "I am to blame".) But the more I look at it, the more that looks like it was a bad idea. We've both come up with some really clever tricks along that road, but the problem is that a user needs to be clever to understand and/or correctly use those tricks. And that's what I'm concerned about. So unless you have ideas for really useful cases where we need to keep doing this (they might exist, I just can't think of any), I think we should stop doing things that take us further into the territory of having movers that can't satisfy detailed balance, and plan to eventually remove the ones we have. (That's not as urgent as finishing 1.0, but it should be on the list.)

We can still re-use tools from proper path sampling for bootstrapping, but it seems dangerous to create new classes that can't be used in path sampling. I'm picturing a future bootstrapping like this: bootstrapping consists of the same start as single replica (choose based on current ensemble), then within each ensemble there's a sequential of shooting move followed by ensemble hop.

This doesn't set the replica IDs, but I think the trick there is to create a little function called `bootstrapping.get_initial_sampleset(storage)` (or maybe takes a list of `mcsteps`?) which loops over the steps and adds to the replica ID each time a new ensemble is encountered.

To me, this is closer to the spirit of what bootstrapping is doing (the change of replica ID always seemed a necessary evil, and as you well know, it is the source of almost all the frustration in implementing bootstrapping). Moving that to an analysis stage cleans it up a lot.

> Like using plus and times with numbers they are very different but are applied the same way.

I would say the better analogy is using plus/times between lists and number in Python. First off, that `[1, 2, 3]*2 == [1, 2, 3, 1, 2, 3]` is not entirely obvious (why not `[2, 4, 6]`?). Second, `1 + [1, 2, 3]` (very reasonably) doesn't work, even though `[1] + [1, 2, 3]` does. Some things can be combined; some can't. It requires a good knowledge of the tool to understand which combinations are allowed and which aren't. And since path movers and detailed balance are a lot more complicated to the average human than lists/numbers, I think we should err toward something less likely to cause mistakes -- especially since, unlike Python, we don't have an easy mechanism to raise an error if an combination is made which is illegal in some particular context.

> I think you came up with the way to use bootstrapping. So, do I understand that you think this is not a good idea now and we should have a clear distinction between sampling and other trajectory generation?

Repeating for emphasis: ***YES.*** Both that it was me who proposed the direction originally, and that I now think it was a bad idea. At the very least, any moves that can not be valid for sampling should carry a flag so that some user who tries to out-clever the laws of statistics by using them can be warned "it appears your plan does not satisfy detailed balance -- are you sure you want to do this?"

> Especially the Generator Idea is still in there which I need to remove and doubles the number of classes.

I noticed this was part of the problem; I'm glad you agree that this needs to be changed. I was going to suggest not having `Generator` in the name. I think having a class variable (my preference) or having them inherit from some abstract class seems like a better way to make it possible for other parts of code to check this.

The conceptual difference is very important: however, I'm not sure when that needs to be checked computationally.